core/src/main/java/com/linecorp/armeria/internal/server/annotation/FileAggregatedMultipart.java

```diff
@@ -48,43 +51,45 @@ ListMultimap<String, String> params() {
-                return resolveTmpFile(multipartUploadsLocation.resolve("incomplete"),
-                                      blockingExecutorService)
-                        .thenComposeAsync(
-                                path -> bodyPart
-                                        .writeTo(path)
-                                        .thenCompose(ignore -> moveFile(
-                                                path,
-                                                multipartUploadsLocation.resolve("complete"),
-                                                blockingExecutorService))
-                                        .thenApply(completePath -> Maps.<String, Object>immutableEntry(
-                                                bodyPart.name(), completePath)),
-                                ctx.eventLoop());
+            final String name = bodyPart.name();
+            final String filename = bodyPart.filename();
+            final EventLoop eventLoop = ctx.eventLoop();
+
+            if (filename != null) {
+                final Path incompleteDir = destination.resolve("incomplete");
+                final ScheduledExecutorService executor = ctx.blockingTaskExecutor().withoutContext();
+
+                return resolveTmpFile(incompleteDir, filename, executor).thenComposeAsync(path -> {
+                    return bodyPart.writeTo(path).thenCompose(ignore -> {
+                        final Path completeDir = destination.resolve("complete");
+                        return moveFile(path, completeDir, executor);
+                    }).thenApply(completePath -> MultipartFile.of(name, filename, completePath.toFile()));
+                }, eventLoop);
             }
-            return bodyPart.aggregate()
-                           .thenApply(aggregatedBodyPart -> Maps.<String, Object>immutableEntry(
-                                   bodyPart.name(), aggregatedBodyPart.contentUtf8()));
-        }).thenApply(result -> {
+
+            return bodyPart.aggregate(eventLoop).thenApply(aggregatedBodyPart -> {
+                return Maps.<String, Object>immutableEntry(name, aggregatedBodyPart.contentUtf8());
+            });
+        }).thenApply(results -> {
             final ImmutableListMultimap.Builder<String, String> params = ImmutableListMultimap.builder();
-            final ImmutableListMultimap.Builder<String, Path> files =
+            final ImmutableListMultimap.Builder<String, MultipartFile> files =
                     ImmutableListMultimap.builder();
-            for (Entry<String, Object> entry : result) {
-                final Object value = entry.getValue();
-                if (value instanceof Path) {
-                    files.put(entry.getKey(), (Path) value);
+            for (Object result : results) {
+                if (result instanceof MultipartFile) {
+                    final MultipartFile multipartFile = (MultipartFile) result;
+                    files.put(multipartFile.name(), multipartFile);
```

Comment:
I just realized that actually `name` is an optional directive (and a NPE will be thrown if it is `null`)

This is actually unrelated to this PR, but might be worth noting/creating a separate issue 😅